app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte

```diff
@@ -0,0 +1,252 @@
+<script lang="ts">
+  import AppPage from "../../../../../app_page.svelte"
+  import FormContainer from "$lib/utils/form_container.svelte"
+  import FormElement from "$lib/utils/form_element.svelte"
+  import FormList from "$lib/utils/form_list.svelte"
+  import { client } from "$lib/api_client"
+  import { page } from "$app/stores"
+  import { goto } from "$app/navigation"
+  import { KilnError, createKilnError } from "$lib/utils/error_handlers"
+
+  // Form fields
+  let name = ""
+  let server_url = ""
+  let description = ""
+
+  // Headers as array of key/value pairs
+  interface HeaderPair {
+    key: string
+    value: string
+    placeholder: string | null
+  }
+
+  let headers: HeaderPair[] = []
+
+  // Form state
+  let error: KilnError | null = null
+  let submitting = false
+
+  // Populate fields from parent page state if provided (only if fields are empty)
+  $: if ($page.state) {
+    const state = $page.state as {
+      name?: string
+      description?: string
+      server_url?: string
+      headers?: HeaderPair[]
+    }
+
+    if (state.name && !name) {
+      name = state.name
+    }
+
+    if (state.description && !description) {
+      description = state.description
+    }
+
+    if (state.server_url && !server_url) {
+      server_url = state.server_url
+    }
+
+    if (state.headers && headers.length === 0) {
+      headers = [...state.headers]
+    }
+  }
+
+  function buildHeadersObject(): Record<string, string> {
+    const headersObj: Record<string, string> = {}
+
+    for (const header of headers) {
+      if (header.key.trim() && header.value.trim()) {
+        headersObj[header.key.trim()] = header.value.trim()
+      }
+    }
+
+    return headersObj
+  }
+
+  async function connect_remote_mcp() {
+    try {
+      error = null
+      submitting = true
+
+      // Validate required fields
+      if (!name.trim()) {
+        throw new Error("Name is required")
```

Comment:
Always do validation logic to the server. The client can always be bypassed. The server owns checking data is valid.